### PR TITLE
Simplify backend mock access for unittests

### PR DESCRIPTION
**Pull request by godexsoft**
- Remove the need for dynamic_cast everywhere
- Refactor impacted unittest code for readability and conciseness
- Add a setRange(min, max) member function in BackendInterface (to be used for testing purposes)

### CODE REVIEW

File: src/data/BackendFactory.h [hunk 1]
```diff
@@ -55,10 +55,8 @@ make_Backend(util::Config const& config)
         throw std::runtime_error("Invalid database type");
 
     auto const rng = backend->hardFetchLedgerRangeNoThrow();
-    if (rng) {
-        backend->updateRange(rng->minSequence);
-        backend->updateRange(rng->maxSequence);
-    }
+    if (rng)
+        backend->setRange(rng->minSequence, rng->maxSequence);
```
Comment (cindyyan317) [inline src/data/BackendFactory.h:59]: 👍

File: src/data/BackendInterface.h [hunk 2]
```diff
@@ -191,6 +191,17 @@ class BackendInterface {
     void
     updateRange(uint32_t newMax);
 
+    /**
+     * @brief Sets the range of sequences that are stored in the DB.
+     *
+     * Warning: Use responsibly (for testing purposes).
```
Comment (cindyyan317) [inline src/data/BackendInterface.h]: can remove this comment now

File: unittests/SubscriptionManagerTests.cpp [hunk 3]
```diff
@@ -190,19 +192,16 @@ TEST_F(SubscriptionManagerSimpleBackendTest, ReportCurrentSubscriber)
...
-    mockBackendPtr->updateRange(10);  // min
-    mockBackendPtr->updateRange(30);  // max
+    backend->setRange(10, 30);
     boost::asio::io_context ctx;
     auto ledgerinfo = CreateLedgerInfo(LEDGERHASH, 30);
     // mock fetchLedgerBySequence return this ledger
-    ON_CALL(*rawBackendPtr, fetchLedgerBySequence).WillByDefault(Return(ledgerinfo));
-    EXPECT_CALL(*rawBackendPtr, fetchLedgerBySequence).Times(1);
+    ON_CALL(*backend, fetchLedgerBySequence).WillByDefault(Return(ledgerinfo));
+    EXPECT_CALL(*backend, fetchLedgerBySequence).Times(1);
```
Comment (kuznetsss) [inline unittests/SubscriptionManagerTests.cpp:200]: Maybe if you are already changing this you could also change `ON_CALL(); EXPECT_CALL().Times(1)` to `EXPECT_CALL()`?
Not important but would look nicer if you can be bothered 🙂

File: src/data/BackendInterface.h [hunk 4]
```diff
@@ -191,6 +191,15 @@ class BackendInterface {
     void
     updateRange(uint32_t newMax);
 
+    /**
+     * @brief Sets the range of sequences that are stored in the DB.
+     *
+     * @param min The new minimum sequence available
+     * @param max The new maximum sequence available
```
Comment (kuznetsss) [inline src/data/BackendInterface.h:198]: Need to add the `force` parameter description

**Review by cindyyan317 — DISMISSED** [review]
LGTM

File: unittests/SubscriptionManagerTests.cpp (hunk 3 above)
Comment (godexsoft) [reply unittests/SubscriptionManagerTests.cpp:200]: This is more than 500 places to change by the look of things.. a quick regex did not help so i'm gonna leave this for another time 👍

**Review by cindyyan317 — APPROVED** [review]

**Review by kuznetsss — APPROVED** [review]

**Pull request merged**